lib/web_ui/lib/src/engine/canvas_pool.dart

```diff
@@ -0,0 +1,761 @@
+          // alpha for the paint the path is painted in addition to the shadow,
+          // which is undesirable.
+          context.save();
+          context.translate(shadow.offsetX, shadow.offsetY);
+          context.filter = _maskFilterToCss(
+              ui.MaskFilter.blur(ui.BlurStyle.normal, shadow.blur));
+          context.strokeStyle = '';
+          context.fillStyle = shadow.color.toCssString();
+          _runPath(context, path);
+          context.fill();
+          context.restore();
+        } else {
+          // TODO(het): We fill the path with this paint, then later we clip
+          // by the same path and fill it with a fully opaque color (we know
+          // the color is fully opaque because `transparentOccluder` is false.
+          // However, due to anti-aliasing of the clip, a few pixels of the
+          // path we are about to paint may still be visible after we fill with
+          // the opaque occluder. For that reason, we fill with the shadow color,
+          // and set the shadow color to fully opaque. This way, the visible
+          // pixels are less opaque and less noticeable.
+          context.save();
+          context.filter = 'none';
+          context.strokeStyle = '';
+          context.fillStyle = shadow.color.toCssString();
+          context.shadowBlur = shadow.blur;
+          context.shadowColor = shadow.color.withAlpha(0xff).toCssString();
+          context.shadowOffsetX = shadow.offsetX;
+          context.shadowOffsetY = shadow.offsetY;
+          _runPath(context, path);
+          context.fill();
+          context.restore();
+        }
+      }
+    }
+  }
+
+  void dispose() {
+    // Webkit has a threshold for the amount of canvas pixels an app can
+    // allocate. Even though our canvases are being garbage-collected as
+    // expected when we don't need them, Webkit keeps track of their sizes
+    // towards the threshold. Setting width and height to zero tricks Webkit
+    // into thinking that this canvas has a zero size so it doesn't count it
+    // towards the threshold.
+    if (browserEngine == BrowserEngine.webkit && _canvas != null) {
+      _canvas.width = _canvas.height = 0;
+    }
+    _clearPool();
+  }
+
+  void _clearPool() {
+    if (_pool != null) {
+      for (html.CanvasElement c in _pool) {
+        if (browserEngine == BrowserEngine.webkit) {
+          c.width = c.height = 0;
+        }
+        c.remove();
+      }
+    }
+    _pool = null;
+  }
+}
+
+// Optimizes applying paint parameters to html canvas.
+//
+// See https://www.w3.org/TR/2dcontext/ for defaults used in this class
+// to initialize current values.
+//
+class ContextStateHandle {
+  html.CanvasRenderingContext2D context;
+  ContextStateHandle(this.context);
+  ui.BlendMode _currentBlendMode = ui.BlendMode.srcOver;
+  ui.StrokeCap _currentStrokeCap = ui.StrokeCap.butt;
+  ui.StrokeJoin _currentStrokeJoin = ui.StrokeJoin.miter;
+  Object _currentFillStyle;
+  Object _currentStrokeStyle;
```

Comment:
Can you please leave a comment explaining why the above two fields need to be `Object`?